README.adoc

```diff
@@ -0,0 +1,154 @@
+
+`org.fedoraproject.Anaconda.Addons.HelloWorld.service`::
+A `systemd` unit file that describes the service to be started. Note that the file is not actually interpreted by `systemd`, so some syntax features will not work! Instead, a process that sets up the D-Bus reads this.
+
+`org.fedoraproject.Anaconda.Addons.HelloWorld.sh`::
+A shell script that sets up environment for `python3` and runs the code in `dbus_service/__main__.py`.
+
+The name of the directory is not required by any rule, but make sure to adjust the `Makefile` if you change the name.
+
+== Makefile
+
+The `Makefile` provided with this addon is very basic. It copies files to their respective paths, and then creates an updates image that contains these files.
+
+The paths for the various types of files encoded in the `Makefile` are required by Anaconda. If you put your own files anywhere else, the addon will not work.
+
+Note that the updates image uses the same format as `initrd` images, but is handled by Anaconda, so caveats apply. In particular, addon images cannot be concatenated like `initrd` ones - you will have to unpack these, merge the trees, and re-compress.
+
+== Addon code directory
+
+All Python code of the addon should be placed in an `<addon_name>` directory. Anaconda addons generally use for this directory a naming scheme that concatenates hiararchically nested identifiers into an URL-like string. You can read `org_fedora_hello_world` as the three components `org`, `fedora`, `hello_world`. You should create your own namespace for your addon, eg. `com_example_widgets`.
+
+In the Python code, the name of this directory corresponds to where your modules are located: `from org_fedora_hello_world.foo.bar import baz`.
+
+In this example addon, this directory contains a single file:
+
+`constants.py`::
+This file contains constants needed by both the service and the user interface code.
+
+Other files shared by both interface and service can go here too, or have their own directory. This part of the tree is not accessed by anything else that your addon's code, so you are free to make up your own rules.
+
+== D-Bus service code
+
+Code in the `<addon_name>/dbus_service` is what runs as the service. The path to this directory is not required by any rule, but make sure to adjust the `Makefile` and configuration in `dbus_data` to match. Additionally, the need to share code between the service and UI means you need to have the directory with the service code somewhere under the `<addon_name>` directory.
+
+The files found here are:
+
+`kickstart.py`::
+Implements classes needed to handle Kickstart data: Parse Kickstart text into internal data structures, and vice versa.
+
+`hello_world.py`::
+Implements the class that represents the service. Binds together the whole service.
+
+`hello_world_interface.py`::
+Implements an interface for the service class. Thanks to the `dasbus` library, this then automagically becomes the actual D-Bus interface.
```

Comment:
I think we should say dasbus is a DBus library or bindings. Also we generally use "DBus" everywhere as far as I can tell from the codebase, at least in Anaconda.

Reply:
Most of the world outside Red Hat calls it D-Bus with a dash, though. So then it becomes the question - which kind of consistency is more important?

Reply:
IIRC, agreement at the team meeting was it doesn't matter that much. Keeping it as D-Bus.